### PR TITLE
(PATCH 2) Cleanup logging

### PR DESCRIPTION
**Pull request by zorkian**
This is unfortunately cluttered with the delta from the gocheck patch. I am not sure how to present it more cleanly. This diff just removes the deprecated `Logger` and also consolidates all logging to use a global log object which is more standard and also more easily integrated with environments like Dropbox's.

### CODE REVIEW

**Comment by optiopay-backend-helper** [comment]
Pull request seem to be stale, assigning @PragmaticCypher as the responsible developer.

**Comment by husio** [comment]
It's good idea to finally remove `Log` attribute, but other changes to logging are not valid. You're proposing global mutable logger, which has at least two problems here:
* it's not thread safe, because you can read/write global `log`
* every time you call `Dial`, you're overwriting configuration for other loggers in process

As for changing to the test suite, I think this is a bit of personal preference wheather you like using external testing libraries or not. I can see that you put effort into making tests better, but I am sorry to say that with current codebase I do not see any gain in using extra testing packages.

**Comment by zorkian** [comment]
It's of course your call if you accept it or not to your repository. I have other major changes coming to rewrite the connection logic to use connection pooling, abstract out metadata handling, rework mutex usage etc., but I want better testing functionality to feel comfortable with them. I also need global log object -- it's easy to make thread safe, of course.

If you can't accept improvements, I will fork the client to Dropbox's repository and it will be hard to upstream fixes as we will move quickly. I feel that would be a shame as this client is really good and I don't want to fracture the open source client ecosystem even more. Kafka doesn't have many good clients as it is.

What do you think? Is there a way to make this work?

**Comment by alpe** [comment]
@zorkian we appreciate your contributions very much but please understand that more people are using this project/ working on it. @husio made good points that need to be addressed. So I'd recommend that you start describing the problem you want to solve first! Everybody can then contribute with ideas and we may have alternative solutions to discuss. I'd prefer a collaborative way to improve this project. 
If github issues won't work for you we could setup a mailing list.

**Comment by zorkian** [comment]
You can read about why you should use Gocheck here: https://labix.org/gocheck

I also pasted the output of using this library -- it's much more readable (particularly now that it prints logging output), it does test timing in verbose mode, you can select tests/suites to work with, etc. The list of advantages is quite long and it will improve the test suite in general. Even if we don't use those features today, we can start to build tests that do now.

The reason for the other improvements is pretty straightforward:

* Mutex handling in the library as-is is not according to Go best practices (use of lock + defer unlock) and, as per the other issue we opened, we've seen it deadlock in high-load situations. We've tried tracking it down to a root cause to no avail, so I think a good first step would be to make the lock usage safer.

* Metadata is currently handled across several areas of the Broker and the `broker.go` file is too long already. It's better to have smaller files with more well contained functionality and Metadata is an obvious one to break out. This is just a code cleanup thing, there is no "imminent need" to do this. (It is helpful to achieve the locking goal of the first issue, though, as the Metadata code does some locking.)

* Finally, the major issue is that throughput in our use case is very hampered by the fact that the client only makes one connection per broker. Kafka will only process one request at a time from a given connection, so the client needs to be able to create multiple connections to each broker to support parallelism. This change is quite involved and is the root cause of why I wanted to do the above cleanups/improvements.

You can see the code for this refactor at: https://github.com/zorkian/kafka/commit/7b13d09afdd8fe99981e17b7b10bff2a67707c05

In short, I've basically performed a major rewrite of the client to address performance needs motivated by limitations in the Kafka protocol to address scaling issues we're having at Dropbox. We definitely think that this library has the interface that we want (it's very idiomatic Go and simple), but we're not tied to any particular implementation.

I'd love to see if this can be upstreamed and I'm definitely willing to put work in to do it -- but it will probably have to be a bit roundabout. I have users/projects right now that are blocked on this performance/hang issue, so I will unblock them internally and separately work with you on getting the improvements upstream in some form or another. Does that work?

**Comment by husio** [comment]
> You can read about why you should use Gocheck here: https://labix.org/gocheck

As I said, this is opinion. You agree with it and you like the library. That's great because if it helps you, then you should definitely use it. But I think that introducing such library to this project right now would bring more harm then help.

> Mutex handling in the library as-is is not according to Go best practices (use of lock + defer unlock) and, as per the other issue we opened, we've seen it deadlock in high-load situations.

I don't agree with statement that using mutex is always bad practice. In case of this library, it allowed us to iterate fast and provide easy to use interface. It might be that this is not solution that produce the fastest interface, but in current state I cannot imagine any other way it could be done.

> Finally, the major issue is that throughput in our use case is very hampered by the fact that the client only makes one connection per broker.

This is known "issue". In optiopay, we solve it by creating separate broker per client.

---

Now that we use this library on production for some time, we also notice that it could have been modeled differently and sometimes it solves the wrong problem (mentioned broker). I think it is a good idea to introduce connection pool, but this must be separate implementation that will not break existing interfaces.

I understand that you have urgent need to solve your problems and you see breaking changes as the only option, but I cannot break everyone else code because of this. Even if breaking change is the only solution, this should be discussed first so that everyone can write down his thoughts and give it enough time to think through. Otherwise you spend time on writing code that may be never accepted.

To summarize this, there is need to change things and to make this library better, but this pull request is not the right way to do this. 
As @alpe mentioned, we will be happy to work on solution that everyone will agree with, but I understand if you think that fork and applying your changes is better.
